pkg/compression/compression.go

```diff
@@ -43,24 +78,44 @@ var compressionAlgos = map[string]struct {
 	"gzip":  {[]byte{0x1F, 0x8B, 0x08}, GzipDecompressor},                 // gzip (RFC 1952)
 	"bzip2": {[]byte{0x42, 0x5A, 0x68}, Bzip2Decompressor},                // bzip2 (decompress.c:BZ2_decompress)
 	"xz":    {[]byte{0xFD, 0x37, 0x7A, 0x58, 0x5A, 0x00}, XzDecompressor}, // xz (/usr/share/doc/xz/xz-file-format.txt)
+	"zstd":  {[]byte{0x28, 0xb5, 0x2f, 0xfd}, ZstdDecompressor},           // zstd (http://www.zstd.net)
+}
+
+// compressors maps an algorithm to its compression function
+var compressors = map[string]CompressorFunc{
+	"gzip":  GzipCompressor,
+	"bzip2": Bzip2Compressor,
+	"xz":    XzCompressor,
+	"zstd":  ZstdCompressor,
+}
+
+// GetCompressor returns the compressor by its name
+func GetCompressor(name string) (CompressorFunc, error) {
+	c, found := compressors[name]
+	if !found {
+		return nil, fmt.Errorf("cannot find compressor for '%s'", name)
+	}
+	return c, nil
 }
 
 // DetectCompression returns a DecompressorFunc if the input is recognized as a compressed format, nil otherwise.
 // Because it consumes the start of input, other consumers must use the returned io.Reader instead to also read from the beginning.
-func DetectCompression(input io.Reader) (DecompressorFunc, io.Reader, error) {
+func DetectCompression(input io.Reader) (string, DecompressorFunc, io.Reader, error) {
```

Comment:
This breaks API, and would require a major version bump. (Yes, dealing with that is annoying.)

Please turn this into something like a `DetectCompressionFormat` that returns the above-mentioned enum, and a compatibility wrapper.

(I’m still a bit hoping that the recompression can just not be implemented instead :) )

Reply:
(It would also be worth considering deprecating the public `pkg/compression` package and intending to make it `internal/compression` in 3.0 — but we should still keep the API in the meantime.

OTOH the package is, in principle, reusable and potentially useful outside of c/image; yet again, I can’t remember any external user.

I don’t have a strong opinion on this, and it’s obviously not directly related to this PR.)